### PR TITLE
Login backend 1

### PR DESCRIPTION
**Pull request by ghost**
### Summary
* Add retrofit call to REST API to RootActivity and LoginPresenter login
* ILoginView updated to manage REST API validation

### Card
https://trello.com/c/ATHZ1i6C/26-u2-backend-login

### Screen Recordings
loginAPI

### CODE REVIEW

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginFragment.java [hunk 1]
```diff
@@ -109,16 +111,22 @@ private void saveLoginDataOnSharedPreferences(String email, String password) {
...
     @Override
     public void onServerError() {
-        //TODO: toast message "Sorry, (bla bla) try again later"
+        Toast toast = Toast.makeText(getContext(), getString(
+                R.string.fragment_login_server_error_toast), Toast.LENGTH_LONG);
+        toast.show();
     }
 
     @Override
     public void onInvalidCallError() {
```
Comment (ghost) [inline app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginFragment.java:127]: Maybe we can check network connection before the toast. The current message was only for testing the different callbacks

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -19,6 +19,7 @@
                 <action android:name="android.intent.action.MAIN" />
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
+
```
Comment (lmiotti) [inline app/src/main/AndroidManifest.xml:22]: Remove this line it's not necessary

File: app/src/main/java/ar/com/wolox/android/example/model/User.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+   @SerializedName("email") private String email;
+   @SerializedName("password") private String password;
+   @SerializedName("picture") private String picture;
+   @SerializedName("cover") private String cover;
+   @SerializedName("description") private String description;
+   @SerializedName("location") private String location;
+   @SerializedName("name") private String name;
+   @SerializedName("phone") private String phone;
+
+    public int getId() {
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/model/User.java:25]: We have to add Nullabilities for this methods. Try to use @Nullable or @Nonnull . The same for every method.

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (ghost) [reply app/src/main/AndroidManifest.xml:22]: I just follow the convention from the other PR corrections: the xml tags that have childs have and empty line above and below. For example: \<application> and \<manifest> has both lines.

But I can even remove the line if you want.

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginFragment.java [hunk 4]
```diff
@@ -106,4 +108,25 @@ private void saveLoginDataOnSharedPreferences(String email, String password) {
...
+    @Override
+    public void onInvalidLoginData() {
+        Toast toast = Toast.makeText(getContext(), getString(
+                R.string.fragment_login_invalid_email_or_password_toast), Toast.LENGTH_LONG);
+        toast.show();
+    }
+
+    @Override
+    public void onServerError() {
+        Toast toast = Toast.makeText(getContext(), getString(
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginFragment.java]: This is code ok, however I would like to use some dependency injection here. Can you change it?

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginFragment.java [hunk 5]
```diff
@@ -106,4 +108,25 @@ private void saveLoginDataOnSharedPreferences(String email, String password) {
...
+                R.string.fragment_login_server_error_toast), Toast.LENGTH_LONG);
+        toast.show();
+    }
+
+    @Override
+    public void onInvalidCallError() {
+        Toast toast = Toast.makeText(getContext(), getString(
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginFragment.java]: Same as above

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java [hunk 6]
```diff
@@ -12,38 +21,64 @@
...
-        boolean isValidEmail = validateEmail(email);
-        boolean isValidPassword = validatePassword(password);
+    void onLoginValidation(String email, String password, String dbBaseUrl) {
+        boolean isValidEmailFormat = validateEmailFormat(email);
+        boolean isValidPasswordFormat = validatePasswordFormat(password);
 
-        if (isValidEmail && isValidPassword) {
-            getView().onValidatedLogin();
+        if (isValidEmailFormat && isValidPasswordFormat) {
+            Retrofit retrofit = new Retrofit.Builder()
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java]: It's not really necessary to do this, we can Inject retrofit services as I suggested to do with ToastFactory

File: app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java [hunk 7]
```diff
@@ -15,25 +25,70 @@
...
         } else {
-            intent = new Intent(this, LoginActivity.class);
+           validateLoginWithDatabase(email, password);
         }
+    }
 
-        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
-        startActivity(intent);
+    private void validateLoginWithDatabase(String email, String password) {
+        Retrofit retrofit = new Retrofit.Builder()
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java]: Same as above

File: app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java [hunk 8]
```diff
@@ -15,25 +25,70 @@
...
+                .baseUrl(getString(R.string.rest_api_base_url))
+                .addConverterFactory(GsonConverterFactory.create())
+                .build();
+        UserService userService = retrofit.create(UserService.class);
+        Call<List<User>> userCall = userService.getUserByEmail(email);
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java]: When injecting retrofitservices we could change this 3 lines in one. Something like
mRetrofitservices.create(UserService.class).getUserByEmail(email).enqueue(.....)
would look better

File: app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java [hunk 9]
```diff
@@ -15,25 +25,70 @@
...
+        userCall.enqueue(new NetworkCallback<List<User>>() {
+            @Override
+            public void onResponseSuccessful(@Nullable List<User> users) {
+                Intent intent;
+
+                if (users != null && !users.isEmpty() && users.get(0).getPassword().
+                        equals(password)) {
+                    intent = new Intent(getApplicationContext(), HomeActivity.class);
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java]: Create two functions goToHomeScreen and goToLoginScreen. I don't like the idea of using intents inside request.

**Review by lmiotti — CHANGES_REQUESTED** [review]
The main idea it's ok. We have to made same change although.

File: app/src/main/java/ar/com/wolox/android/example/model/User.java (hunk 3 above)
Comment (ghost) [reply app/src/main/java/ar/com/wolox/android/example/model/User.java:25]: ✔ **Done**

But assuming email, username and password can't be null and the other fields can be null. I don't know if that assumption is true...

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginFragment.java (hunk 4 above)
Comment (ghost) [reply app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginFragment.java]: ✔ **Done**

Now I understand and love injection :smile_cat:

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginFragment.java (hunk 5 above)
Comment (ghost) [reply app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginFragment.java]: ✔ **Done**

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java (hunk 6 above)
Comment (ghost) [reply app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java]: ✔ **Done**

File: app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java (hunk 7 above)
Comment (ghost) [reply app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java]: ✔ **Done**

File: app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java (hunk 8 above)
Comment (ghost) [reply app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java]: ✔ **Done**

Yes, looks really better :grinning:

File: app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java (hunk 9 above)
Comment (ghost) [reply app/src/main/java/ar/com/wolox/android/example/ui/example/root/RootActivity.java]: ✔ **Done**

I thoughed the same, I had even done the launchLoginActivity() function, but then I forgot to do the same for HomeActivity :disappointed:

**Comment by ghost** [comment]
> The main idea it's ok. We have to made same change although.

Changes done! I learned a lot, specially about Dependency Injection :smiley:

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java [hunk 10]
```diff
@@ -1,49 +1,80 @@
...
  *
  */
-class LoginPresenter extends BasePresenter<ILoginView> {
+public class LoginPresenter extends BasePresenter<ILoginView> {
+
+    @Inject RetrofitServices mRetrofitService;
 
     @Inject LoginPresenter() {}
 
     void onLoginValidation(String email, String password) {
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java]: We can add Nullabilities here too. We know that email and password will be not null so @Nonnull string email and @Nonnull String password can be included as parameters

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java [hunk 11]
```diff
@@ -1,49 +1,80 @@
...
+                public void onCallFailure(Throwable throwable) {
+                    getView().onInvalidCallError();
+                }
+            });
         }
     }
 
-    private boolean validateEmail(String email) {
-        boolean isValidEmail = false;
+    private boolean validateEmailFormat(String email) {
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java]: Same as above

**Review by lmiotti — CHANGES_REQUESTED** [review]
Just two minor changes are pending

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java (hunk 10 above)
Comment (ghost) [reply app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java]: ✔ **Done**

File: app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java (hunk 11 above)
Comment (ghost) [reply app/src/main/java/ar/com/wolox/android/example/ui/example/login/LoginPresenter.java]: ✔ **Done**

Added on validatePasswordFormat(String password) too

**Comment by ghost** [comment]
> Just two minor changes are pending

Changes done

**Review by lmiotti — APPROVED** [review]
tom cruise

**Pull request merged**